Core/Services/StitchCoreRemoteMongoDBService/Sources/StitchCoreRemoteMongoDBService/Sync/CoreDocumentSynchronizationConfig.swift

```diff
@@ -0,0 +1,338 @@
+            try! filter.merge([Config.CodingKeys.isStale.rawValue: true])
+            let count = try! docsColl.count(filter)
+            return count == 1
+        }
+        set(value) {
+            docLock.writeLock()
+            defer { docLock.unlock() }
+            let _ = try! docsColl.updateOne(
+                filter: docFilter(forNamespace: namespace, withDocumentId: documentId),
+                update: ["$set": [Config.CodingKeys.isStale.rawValue: value] as Document])
+            self.config.isStale = value
+        }
+    }
+
+    /// Whether or not this document has been paused due to an error.
+    var isPaused: Bool {
+        get {
+            docLock.readLock()
+            defer { docLock.unlock() }
+            return config.isPaused
+        }
+        set(value) {
+            docLock.writeLock()
+            defer { docLock.unlock() }
+            let _ = try! docsColl.updateOne(
+                filter: docFilter(forNamespace: namespace,
+                                  withDocumentId: documentId),
+                update: [ "$set": [ Config.CodingKeys.isPaused.rawValue : value ] as Document
+                ])
+            config.isPaused = value
+        }
+    }
+
+    /// Whether or not there is a pending write for this document.
+    var hasUncommittedWrites: Bool {
+        get {
+            return lastUncommittedChangeEvent != nil
+        }
+    }
+
+    init(docsColl: MongoCollection<CoreDocumentSynchronization.Config>,
+         namespace: MongoNamespace,
+         documentId: AnyBSONValue) {
+        self.docsColl = docsColl
+        self.config = Config.init(namespace: namespace,
+                                  documentId: HashableBSONValue.init(documentId),
+                                  lastUncommittedChangeEvent: nil,
+                                  lastResolution: -1,
+                                  lastKnownRemoteVersion: nil,
+                                  isStale: false,
+                                  isPaused: false)
+    }
+
+    init(docsColl: MongoCollection<CoreDocumentSynchronization.Config>,
+         config: inout Config) {
+        self.docsColl = docsColl
+        self.config = config
+    }
+
+    /**
+     Sets that there are some pending writes that occurred at a time for an associated
+     locally emitted change event. This variant maintains the last version set.
+
+     - parameter atTime: the time at which the write occurred.
+     - parameter changeEvent: the description of the write/change.
+     */
+    mutating func setSomePendingWrites(atTime: TimeInterval,
+                                       changeEvent: ChangeEvent<Document>) {
+        // if we were frozen
+        if (isPaused) {
+            // unfreeze the document due to the local write
+            isPaused = false
+            // and now the unfrozen document is now stale
+            isStale = true
+        }
+
+        docLock.writeLock()
```

Comment:
[major] shouldn't we acquire write lock before changing paused/stale status?

Reply:
I ported over what happens on Android and assumed we had a reason for doing it this way. @edaniels?

Reply:
This is handled implicitly in the setters. I can make it explicit, however this is not a design flaw.